CHANGELOG.md

```diff
@@ -27,6 +27,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - `StateMachine::new` and `StateMachine::new_with_state` are now const functions
 - Fixed clippy warnings
 - [breaking] Changed guard functions return type from Result<(),_> to Result<bool,_>
+- [breaking] Changed action functions return type from () to Result<NextStateData,_>
+- [breaking] Made guard's self argument immutable
```

Comment:
Hi @ryan-summers ,

Here are some thoughts on disallowing `guards` from mutating the state machine's extended state:

  - Separation of Concerns:
Guards are intended to serve as pure functions that only evaluate conditions. Their primary role is to check whether a transition should occur, not to perform actions or change the state.
  - Predictability and Clarity:
Allowing guards to mutate the state can make the state machine's behaviour less predictable and harder to understand. When guards are pure and side-effect-free, it's easier to reason about the system since the guards only depend on the current state and the event.
  - Avoiding Side Effects:
Guards with side effects can introduce unintended consequences and bugs. For example if guards are allowed to modify the state, the logic will depend on the order in which guards are evaluated. For example [a&&b] and [b&&a] may give different results.  Keeping guards pure ensures that evaluating a guard does not alter the system's state, maintaining a clear separation between decision logic and state modification. 
  - Functional Programming Principles:
Rust encourages functional programming principles, including immutability and pure functions. By adhering to these principles, the codebase can be more robust, testable, and maintainable. Guards that mutate state violate these principles.

What if a guard needs for example,  to communicate  say to an I2C  sensors to measure temperature and it needs mutable access to context to achieve that?

In cases like that it is essential to differentiate between the roles of guards and actions in a state machine. Guards should remain pure and side-effect-free, but actions are associated with transitions, and they  can perform the necessary operations that require mutable access. 

The following strategies  can be used 
  - Pre-Read Data: Read the necessary data (e.g., temperature) before evaluating the guard and store it in a temporary variable.
  - Separate Guard and Action Logic: Ensure that guards only check conditions based on available data, and any necessary state changes or side-effect operations are performed in actions or transition functions.

I hope this will convince to keep the proposed change.

Reply:
The more I think about this the more I like it. Indeed a guard should not have side effects, and you make a good point that a guard needing to i.e. acquire external data could do that and cache the result elsewhere (and there's other workarounds).

Given that actions expose mutable access, I think this all makes sense. Thanks for the discussion.